### ISSUE TITLE
CSRF mismatch in production

### ISSUE THREAD

**Issue by callahad** [mozilla/persona#3473]
Possibly related to move to AWS?
### IRC Log

buro9: I've got a user who cannot login to Persona
buro9: I've pointed them at http://lloyd.github.io/persona-troubleshoot/troubleshoot.html and the only failure is the 'onlogout' says no
buro9: the error message she sees is:
buro9: Action: Verifying email address, Network Info: POST: /wsapi/complete_user_creation, Response Code - 400, Response Text: Bad Request: CSRF violation
buro9: she's tried 3 different email accounts all with the same error (she may well be using the same password on all accounts)
buro9: and she's using IE10 on Win7
buro9: anyone seen this before?
buro9: I had the user try in Firefox (v21 on Win7), and the issue was also present there... 3 email accounts, 2 browsers... the only thing I wondered whether her password was weird (unicode chars?) or whether she was behind some proxy that mangled something in some consistent way
### About 'CSRF violation' errors

Those get thrown on [wsapi.js#L158](https://github.com/mozilla/browserid/blob/dev/lib/wsapi.js#L158). Due to other checks in that function, we know that it's because either `!req.body` or because `req.body.csrf !== req.session.csrf`.

Line 157 spits out a warning into the logs, `logger.warn("CSRF validation failure, token mismatch. got:" + b + " want:" + s);`, which would tell us what the two values are, and at least let us determine if the tokens are making it to the back-end at all.
### Next Steps

@lloyd suspects that this may be related to the move to AWS, so tagging 5-Star for now.

@gene1wood / @jrgm can one of you please grep the logs for the last 24 hours and find occurrences of CSRF validation failures?

@lloyd also notes that we should probably have a statsd counter in there.

**Comment by grafana-dee** [comment]
User explanation:

> I get through to the final stage but when I click on the verification email and enter password I get the following message.

Error message:

```
We are very sorry. There has been an error!

Please close this window and try again.

See more info : https://login.persona.org/verify_email_address?token=jmBjoR5eJo5JEYEC4aKiLK4cfT0ZyFzdRu0TAGpWs5Q3m7cq

    Action: Verifying email address
    Network Info: POST: /wsapi/complete_user_creation 

Response Code - 400

Response Text: Bad Request: CSRF violation

Error Type: Bad Request: CSRF violation
```

I've asked her to complete http://lloyd.github.io/persona-troubleshoot/troubleshoot.html and http://supportdetails.com/ so that we have more info, and the info from those is attached.

![persona1](https://f.cloud.github.com/assets/72949/585855/5169d744-c93f-11e2-86c5-2086512ec6f9.png)
![persona3](https://f.cloud.github.com/assets/72949/585857/51803b6a-c93f-11e2-8302-c997336fdf9d.png)
![persona2](https://f.cloud.github.com/assets/72949/585856/51817fd4-c93f-11e2-87fc-24ce805191f8.png)

I do have contact with the user, and can provide support to her, potentially using Chrome to do remote desktop, or by driving the 100 miles and giving on-site support.

**Comment by grafana-dee** [comment]
She had previously tried using Persona on May 13th and received this error:

> Tried to log in again, this is the text that I got?

```
    Action: Creating Account
    Network Info: POST: /wsapi/stage_user

Response Code - 400

Response Text: Bad Request: CSRF violation

Error Type: Bad Request: CSRF violation 
```

**Comment by jaredhirsch** [comment]
bumped down to 4 stars because this same issue (CSRF problems) occurred before our move to AWS

**Comment by grafana-dee** [comment]
The end user is coming to London next Friday (7th June) and I will have access to her machine (with her present) for about an hour in the late afternoon (BST). If you would like the issue reproduced on cue and with a technical person (me) witnessing and testing scenarios (different browser settings, alternate passwords, debugging and dumping traces, etc) then let me know if that will help and I will arrange it.

**Comment by shane-tomlinson** [comment]
Hi @buro9, my name is Shane Tomlinson, I am one of the developers on the Persona team. I also live in London and can help out next Friday.

**Comment by grafana-dee** [comment]
This is no longer an issue, but I do not know what fixed it.

Back story: The user is no longer coming to London on Friday and I was investigating whether remote support would be possible. As the primary email she wants to use is a new one without access to sensitive info, she reset her password and shared it so that I could try and recreate the issue. I was able to successfully create the account (she was not registered, and I had to create her password) using an up-to-date Firefox 21 on Ubuntu.

The email notification was dispatched, so I asked her to confirm it. She apparently tried to login again and failed on IE, but then copied the link into Firefox and that worked for her... she's activated the account and logged in.

I've had her sign out, and sign in... all fine in Mozilla. She's now happy and there is no issue in her opinion, though it nags as her IE issues were not explained.

I haven't been able to reproduce the issue, only the end user has been able to. Given that, I think it should be closed unless it proves reproducible.
